File: my-solutions/0883-car-fleet/solution.cpp
#include<stack>

class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        vector<std::pair<int, int>> pos_speed;
        for(int i = 0; i < size(position); i++) {
            pos_speed.push_back(std::pair<int, int>(position[i], speed[i]));
        }
        std::sort(pos_speed.begin(), pos_speed.end(), [](const std::pair<int, int> a, const std::pair<int, int> b){
            return a.first > b.first;
        });

        // for(int i = 0; i < size(position); i++) {
        //     cout << pos_speed[i].first << " " << pos_speed[i].second << endl;
        // }

        std::stack<std::pair<int, int>> car_stack;

        car_stack.push(pos_speed[0]);

        for(int i = 1; i < size(pos_speed); i++) {
            float curr_time = float(float(target - pos_speed[i].first) / float(pos_speed[i].second));
            float last_time = float(float(target - car_stack.top().first) / float(car_stack.top().second));

            if(curr_time > last_time) {
                car_stack.push(pos_speed[i]);
            }
        }

        return size(car_stack);
    }
};

File: my-solutions/0883-car-fleet/solution.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        posSpeed = [[position[i], speed[i]] for i in range(len(speed))]
        posSpeed.sort(key = lambda a: a[0], reverse = True)
        stack = deque()
        
        def clash(ps1, ps2):
            if ps1[1] == ps2[1]:
                return ps1[0] == ps2[0]
            else:
                t = float(ps2[0] - ps1[0]) / float(ps1[1] - ps2[1])
                # print(ps1[0], ps1[1], ps2[0], ps2[1])
                x = float(ps1[0]) + float(ps1[1]) * t
                # print(t, x)
                return t >= 0 and x <= target
        
        for i in range(len(position)):
            if i == 0:
                stack.append(i)
                continue
            # print(i, stack)
            # print(posSpeed[i], posSpeed[stack[-1]])
            if clash(posSpeed[i], posSpeed[stack[-1]]):
                continue
            else:
                stack.append(i)
            
        return len(stack)
